src/devices/src/virtio/net/device.rs

```diff
@@ -497,242 +514,144 @@ impl NetEpollHandler {
-                    })
-                } else if !self.tx.rate_limiter.is_blocked()
-                // If the limiter is not blocked, continue transmitting bytes.
-                {
-                    self.process_tx()
-                } else {
-                    Ok(())
-                }
-            }
-            RX_RATE_LIMITER_EVENT => {
-                METRICS.net.rx_event_rate_limiter_count.inc();
-                // Upon rate limiter event, call the rate limiter handler
-                // and restart processing the queue.
-                match self.rx.rate_limiter.event_handler() {
-                    Ok(_) => {
-                        // There might be enough budget now to receive the frame.
-                        self.resume_rx()
-                    }
+        // While limiter is blocked, don't process any more incoming.
+        if self.rx_rate_limiter.is_blocked() {
+            return;
+        }
+
+        if self.rx_deferred_frame
+        // Process a deferred frame first if available. Don't read from tap again
+        // until we manage to receive this deferred frame.
+        {
+            if self.rate_limited_rx_single_frame() {
+                self.rx_deferred_frame = false;
+                match self.process_rx() {
+                    Ok(()) => (),
                     Err(e) => {
+                        error!("{:?}", e);
                         METRICS.net.event_fails.inc();
-                        error!("Failed to get rx rate-limiter event: {:?}", e);
-                        Err(DeviceError::RateLimited(e))
                     }
                 }
-            }
-            TX_RATE_LIMITER_EVENT => {
-                METRICS.net.tx_rate_limiter_event_count.inc();
-                // Upon rate limiter event, call the rate limiter handler
-                // and restart processing the queue.
-                match self.tx.rate_limiter.event_handler() {
-                    Ok(_) => {
-                        // There might be enough budget now to send the frame.
-                        self.process_tx()
-                    }
+            } else if self.rx_deferred_irqs {
+                self.rx_deferred_irqs = false;
+                match self.signal_used_queue() {
+                    Ok(()) => (),
                     Err(e) => {
+                        error!("{:?}", e);
                         METRICS.net.event_fails.inc();
-                        error!("Failed to get tx rate-limiter event: {:?}", e);
-                        Err(DeviceError::RateLimited(e))
                     }
                 }
             }
-            other => Err(DeviceError::UnknownEvent {
-                device: "net",
-                event: other,
-            }),
+        } else {
+            match self.process_rx() {
+                Ok(()) => (),
+                Err(e) => {
+                    error!("{:?}", e);
+                    METRICS.net.event_fails.inc();
+                }
+            }
```

Comment:
There is this identical `match` (where just the object being matched is different) in a lot of places, I suggest we add a simple static function:
```rust
fn report_err(err: &DeviceError) {
    error!("{:?}", e);
    METRICS.net.event_fails.inc();
}
```

and all these `match`es can then become:

```rust
self.process_rx().unwrap_or_else(report_err);
```